File: Languages/Ruby/Basics/conditionals.rb
# Conditionals

# IF statement : Takes a expression/boolean value
# Executes following code only if true

# Generate a random number between 0 and 10 (not included)
num = rand(10)

# Check if the number is even
if num % 2 == 0
  puts num.to_s + " is even"
end

# IF else
# else statement will execute another set of code if the condition is false and if block is not run

num = rand(100)
# Check if number is divisible by 11 or not
if num % 11 == 0
  puts num.to_s + " is divisible by 11"
else
  puts num.to_s + " is not divisible by 11"
end

# Combine conditions
# Using a combination of operators we can have a complex express

# Check if number is greater than 50 and divisible by 2
if num % 2 == 0 and num > 50
  puts "Condition satisfied"
else
  puts "Condition failed"
end

# If Else ladder
# Instead of just using one condition we can have a series of conditions executed after failure of one above
# This construct is called if else ladder

# Check fo divisiblity by 2, if not then 3, if not then 5
if num % 2 == 0
  puts num.to_s + " is divisible by 2"
elsif num % 3 == 0
  puts num.to_s + " is divisible by 3"
elsif num % 5 == 0
  puts num.to_s + " is divisible by 5"
else
  puts "Conditions failed"
end

# Case expressions
# Instead of comparing one variable for different values using a long if elsif ladder we can use case expressions
# It compares the variable/constant to a list of values
# When a match occurs, the corrsponding block of code is run

# A random day generator
day = rand(7) + 1
puts day

# Depending on the number above print the day
print "Today is "

case day
when 1
  puts "Monday"
when 2
  puts "Tuesday"
when 3
  puts "Wednesday"
when 4
  puts "Thursday"
when 5
  puts "Friday"
when 6
  puts "Saturday"
when 7
  puts "Sunday"
else
  puts "Invalid"
end

File: Languages/PHP/Basics/conditionals.php
<?php

    /*
    Conditionals are closely realted to BOolean
    That is because most of the conditions are expressions that evaluate to true or false.
    Which are the two boolean values
    */

    /* A simple price comparison using If
    If Statements that a condition or a combination of them and check if the final value is 1/true/Non-Null or 0/false/Null
    If the result is true then the immediate line/block of code follwoing the If statement will be executed.
    If false then the same code will be skipped
    */
    $price = 30;
    if ($price > 25) {
        echo "Condition met";
    }

    /* IF ELSE example
    On the another hand we can use else statements to handle situations when the if statement returns false.
    It will execute another set of code only if the if statement is false and skipped.
    Note : Single line codes dont need to be enclosed in {} brackets.
    */

    if ($price > 25) 
        echo "Expensive";
    else 
        echo "Cheap";
    

    /* IF ELSE Ladder
    Taking it to the next level we can multiple if else conditions creating a ladder like structure
    If one condition is not met it is compared to the other and the process is repeated until the end
    The final else staement for when all comparisons fail can be skipped too
    Note. There is no space between else and if when they appear together
    */

    
    if ($price > 25) 
        echo "Expensive";
    elseif ($price > 15)
        echo "Affordable";
    else 
        echo "Cheap";

    /* Multiple conditions in same statement
    Combining a host of logical (&&, ||, !, and, or, xor) and relational operators (>,<,==,!=,>=,<=, <>,<=>),
    we can create a complicated conditon
    */

    if( ($price > 20 or $price < 30) && ($price % 2 ==0))
        echo "Price is between 20 and 30, is divisible by 2";
    

    /* Switch statements
    Other popular conditional construct is the Switch statement. 
    Sometimes we need to compare the value of a variable against a list of values
    And for each match we find we want a different block of code to execude.
    This is where switch statements come in handy. 
    */

    $option = 3;

    # An example in IF ELSE Ladder
    if($option == 1)
        echo "option 1";
    else if($option == 2)
        echo "option 2";
    else if($option == 3)
        echo "option 3";
    else if($option == 4)
        echo "option 5";
    else 
        echo "invalid option";

    # Now the above statements are totally but a bit too verbose. 
    # We can do it in a better way using switch

    switch($option){
        case 1: echo "option 1"; 
                break;
        case 2: echo "option 2"; 
                break;
        case 3: echo "option 3"; 
                break;
        case 4: echo "option 4"; 
                break;
        default : echo "invalid option";
    }   
    
    /* Switch basics
    So we pass a single variable, not an expression in switch statement
    It's value is compared to a number of cases and whichever case matches it's corresponding code is executed
    Else we have the default case (optional) for when no cases match
    Also the break statement is there to signify the end of a case's code block
    Without the break staement, case 1 if selected will also select all cases below.
    This is called fallthrough phenomena 
    */

?>
